arch/sim/src/sim/up_timer.c

```diff
@@ -0,0 +1,362 @@
+{
+  struct sigaction act;
+  int ret;
+  sigset_t set;
+
+  act.sa_sigaction = action;
+  sigemptyset(&act.sa_mask);
+  act.sa_flags = SA_SIGINFO;
+
+  sigemptyset(&set);
+  sigaddset(&set, SIGALRM);
+  pthread_sigmask(SIG_UNBLOCK, &set, NULL);
+
+  if ((ret = sigaction(SIGALRM, &act, NULL)) != 0)
+    {
+      fprintf(stderr, "ERROR %d signal handler", ret);
+    }
+
+  return ret;
+}
+
+/****************************************************************************
+ * Name: get_irqstate_from_signal_mask
+ *
+ * Description:
+ *   A helper function that creates the NuttX IRQ mask from the host signal
+ *   mask.
+ *
+ * Input Parameters:
+ *   signal_mask - the signal mask for the current running process on host
+ *
+ * Description:
+ *   Return a bitmask with the current active signals.
+ *
+ ****************************************************************************/
+
+static uint32_t get_irqstate_from_signal_mask(const sigset_t *signal_mask)
+{
+  uint32_t nx_irq_mask = 0;
+
+  for (int bit_index = 0; bit_index < PRIMASK_BIT_SIZE; bit_index++)
+    {
+      if (sigismember(signal_mask, g_host_signal_map[bit_index]))
+        {
+          nx_irq_mask |= (1 << bit_index);
+        }
+    }
+
+  return nx_irq_mask;
+}
+
+/****************************************************************************
+ * Name: get_signal_mask_from_irq_state
+ *
+ * Description:
+ *   A helper function that creates a host signal mask from NuttX IRQ mask.
+ *
+ * Input Parameters:
+ *   restored_mask - (out) buffer where we store the active signals
+ *   irqstate      - the mask with the active signals
+ *
+ ****************************************************************************/
+
+static void get_signal_mask_from_irqstate(sigset_t *restored_mask,
+                                          uint32_t irqstate)
+{
+  if (restored_mask == NULL)
+    {
+      return;
+    }
+
+  for (int bit_index = 0; bit_index < PRIMASK_BIT_SIZE; bit_index++)
+    {
+      if ((irqstate & (1 << bit_index)) != 0)
+        {
+          sigaddset(restored_mask, g_host_signal_map[bit_index]);
+        }
+    }
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_prepare_timer
+ *
+ * Description:
+ *   Blocks all the signals in the current thread so that the new threads
+ *   can inherit a blocked signal mask.
+ *
+ * Returned Value:
+ *   This function returns 0 on success otherwise -1.
+ *
+ * Assumptions/Limitations:
+ *   This function should be called prior to spawning new threads as we
+ *   don't want to end up running nxsched_process_timer on a different
+ *   host thread.
+ *
+ ****************************************************************************/
+
+int up_prepare_timer(void)
```

Comment:
Please merge all related logic into up_oneshot.c. It doesn't make sense that we implment timer twice.

Reply:
I merged all the logic in up_oneshot.c, thanks for the feedback